### PR TITLE
Flag JS Literals and ignore assignments/accesses to invalid props, instead of adding an index

### LINKED ISSUE #25987: `keyof` on types originating from javascript object literals is (nearly) useless
**TypeScript Version:**  3.1.0-dev.201xxxxx

**Search Terms:**

**Code**

In a js file:
```ts
// @ts-check
const x = {
    y: 12
};

/** @type {keyof typeof x} */
const q = "z"; // Assignment is allowed, but is not what was desired!

```

**Expected behavior:**
`keyof` returns the keys the object actually has or is declared with.

**Actual behavior:**
`keyof` returns `string | number`, thanks the the any-index signature shoved into every object literal expression's type in JS.

This came up on `DefinitelyTyped` - strongly typing react's `propTypes` is virtually impossible because all of the key filtering and mapping operations (eg, `Exclude`) just _don't work_ on any JS object literals (unless all of your keys are unique symbols (nobody would do this, I don't think), since those aren't mangled by subtype reduction with the index signature).

cc @mhegazy This is a blocker in rolling out the better React inferences to our JS users.

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #25987

As mentioned in-person with @mhegazy, since our goal with the index signature was to just squelch a class of errors in JS, this accomplishes the same thing without the unfortunate side effect of mangling our inferences. This has the fortunate side-effect that the meaningless index signatures are no longer present in quickinfo and the like.

These accesses are also now marked in `noImplicitAny` mode now - which I'd argue is a correct thing to do. That is the breaking change referenced in the PR labels, however in practice it's unlikely to break much (and when it does it's probably desirable!) as JS codebases using `noImplicitAny` are relatively rare.

### CODE REVIEW

**Comment by brendankenny** [comment]
> As an aside, these accesses could totally be marked in `noImplicitAny` mode now - which I'd argue is a correct thing to do; we can put off deciding to do that, though.

One downside of this change (if I'm understanding it correctly) is that not having `[x: string]: any;` in the interface anymore means that it's going to be more difficult to diagnose what's going on when those accesses are not marked as an error. Right now it's kind of a process of hovering for quickinfo and then realizing, "oh yeah, tsc treats object literals loosely here" :)

Having `keyof`, etc work as expected is definitely an improvement for checked js, but it would be *fantastic* if we're also able to opt into errors for them with `noImplicitAny`.

**Comment by mhegazy** [comment]
Since we are touching this area, i agree with @brendankenny that we should make `--noImplicitAny` work as expected. we have talked about this before but never done it.

**Comment by weswigham** [comment]
@mhegazy Alright, we'll no longer squeltch the errors under `noImplicitAny`. It's up now~

@sandersn I fixed a pair of small problems (like also disabling excess property checks on js literals outside of `noImplicitAny` mode), but it looks like now the changes are _mostly_ pretty minimal. There's a lot of churn of index signatures getting removed and I believe errors from the recent lib update which we'll need to look over tomorrow; but this looks pretty good, IMO.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -9072,6 +9075,34 @@ namespace ts {
...
+         */
+        function isJSLiteralType(type: Type): boolean {
+            if (noImplicitAny) {
+                return false; // Flag is meaningless under `noImplicitAny` mode
+            }
+            // In addition to objects which are directly literals,
+            // * unions where every element is a jsliteral
+            // * intersections where at least one element is a jsliteral
+            // * and instantiable types constrained to a jsliteral
+            // Should all count as literals and not print errors on access or assignment of imaginary properties.
```
Comment (sandersn) [inline src/compiler/checker.ts]: quibble: these properties are not "imaginary". They are "possibly existing" properties.

Seriously, though, I'd say "expando properties". I also don't think that referencing the old implementation helps, so I'd delete that line or at least change it to "This mirrors the behaviour of index signatures".

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -9072,6 +9075,34 @@ namespace ts {
             return type;
         }
 
+        /**
+         * Returns if a type is or consists of a JSLiteral object type
```
Comment (sandersn) [inline src/compiler/checker.ts]: I would move the comment block up here.

**Comment by weswigham** [comment]
I'm going to hold off on merging this until our user/rwc baselines are clean, so we can clearly see its impact on them. AFAIK we're not shipping this in a `3.0.2` (it's technically a break, after all), so this isn't in much of a rush.

**Pull request merged**